File: repos/openrappter/typescript/src/rappids/growth.ts
/**
 * Growth: append a verified body frame, never mint a second identity.
 *
 * A proposal is not organism state. It is a preview with a content-addressed
 * id, and the id is what makes approval mean something: `grow` re-derives the
 * proposal from the organism itself and refuses anything whose id does not
 * match, so an approved preview cannot be swapped for a different payload
 * between the preview and the append.
 *
 * Appending never touches the RAPPID. A creature that grows from baby to
 * Raptor is the same creature the whole way — only a true child gets a new
 * identity, and it gets an explicit parent pointer with it.
 *
 * Mirrored by `python/openrappter/rappids/growth.py`.
 */

import {
  RAPP_EGG_DOMAIN,
  canonicalJson,
  rappHb,
  sha256Hex,
} from './canonical.js';
import { proposeContinuation } from './autocomplete.js';
import { buildDnaPrompt, sonicParameters, writeMidi } from './midi.js';
import {
  appendBodyFrame,
  buildDimensionFrame,
  FRAME_TIME_PATTERN,
  formatFrameTime,
  mediaRef,
  writeDimensionAsset,
} from './store.js';
import {
  contiguousFrameHeight,
  deriveStage,
  deriveStageFromEvidence,
  deriveStats,
  dimensionStates,
  summarize,
} from './stats.js';
import { assertVerified, verifyOrganism } from './verify.js';
import { QuantumRappidError } from './types.js';
import type {
  AssetRecord,
  AssetVerification,
  BodyFrame,
  ContinuationProposal,
  GrowthProposal,
  JsonObject,
  LoadedOrganism,
  MusicalParameters,
  Note,
  QuantumRappidSummary,
  VerificationReport,
} from './types.js';

/** Growth requests this runtime knows how to answer. Anything else is refused. */
export const GROWABLE_DIMENSIONS = ['sonic', 'stats'] as const;
export type GrowableDimension = (typeof GROWABLE_DIMENSIONS)[number];

/**
 * The instant a prediction is drafted against.
 *
 * Frame timestamps are fixed-width, so any value of the right shape produces a
 * frame of the same canonical length — which is what lets a preview quote
 * exact bytes for a frame that has not been written yet.
 */
const PREDICTION_INSTANT = '1970-01-01T00:00:00.000Z';
const STAGE_ORDINAL = { baby: 0, hatchling: 1, raptor: 2 } as const;

export interface PendingGrowth {
  proposal: GrowthProposal;
  frameKind: 'body.dimension';
  /** Bytes that would be written, keyed by the asset path they land on. */
  payloads: Array<{ dimension: string; path: string; bytes: Buffer }>;
  continuation: ContinuationProposal | null;
}

function isGrowable(dimension: string): dimension is GrowableDimension {
  return (GROWABLE_DIMENSIONS as readonly string[]).includes(dimension);
}

/** The prompt and key an organism sings in, recorded if it has a profile. */
export function sonicContext(organism: LoadedOrganism): { params: MusicalParameters; prompt: Note[] } {
  const params = organism.sonic?.musical
    ?? sonicParameters(
      organism.document.rappid,
      organism.traits.birthTraitsMilli,
    );
  const prompt = organism.sonic?.prompt
    ?? buildDnaPrompt(
      organism.document.rappid,
      organism.traits.birthTraitsMilli,
      params,
    );
  return { params, prompt: [...prompt] };
}

function proposalId(body: JsonObject): string {
  return sha256Hex(`quantum-rappid/1:proposal\n${canonicalJson(body)}`);
}

/** A verified asset record for bytes that are in hand but not yet on disk. */
function pendingVerification(
  dimension: string,
  asset: AssetRecord,
  bytes: Buffer,
): AssetVerification {
  const addressHash = rappHb(RAPP_EGG_DOMAIN, bytes);
  return {
    dimension,
    path: asset.path,
    status: 'verified',
    addressSpace: RAPP_EGG_DOMAIN,
    addressHash,
    expectedBytes: asset.bytes,
    actualBytes: asset.bytes,
    expectedSha256: asset.sha256,
    actualSha256: asset.sha256,
    mediaType: asset.mediaType,
  };
}

/**
 * What the organism would look like after this frame, without writing it.
 *
 * The prediction runs the real derivation over a copy rather than a parallel
 * "estimate" path: a preview that is computed differently from the thing it
 * previews is how a habitat starts lying to its operator.
 */
function predict(
  organism: LoadedOrganism,
  report: VerificationReport,
  frame: BodyFrame,
  payloads: ReadonlyArray<{ asset: AssetRecord; bytes: Buffer }>,
): { stats: GrowthProposal['predictedStats']; stage: GrowthProposal['predictedStage'] } {
  const projected: LoadedOrganism = { ...organism, frames: [...organism.frames, frame] };
  const projectedReport: VerificationReport = {
    ...report,
    assets: [
      ...report.assets,
      ...payloads.map(({ asset, bytes }) =>
        pendingVerification(frame.payload.dimension, asset, bytes),
      ),
    ],
    verifiedAddresses: [
      ...new Set([
        ...report.verifiedAddresses,
        ...payloads.map(({ bytes }) =>
          `${RAPP_EGG_DOMAIN}:${rappHb(RAPP_EGG_DOMAIN, bytes)}`,
        ),
      ]),
    ].sort(),
  };
  const dimensions = dimensionStates(projected, projectedReport);
  const stats = deriveStats(projected, projectedReport, dimensions);
  return { stats, stage: deriveStage(stats, dimensions) };
}

/**
 * Build the frame a proposal would append.
 *
 * `createdAt` is an input rather than a clock read, because the frame hash
 * covers it: a proposal id that changed every second could never be approved.
 */
export function projectedStage(
  organism: LoadedOrganism,
  dimension: string,
  report: VerificationReport = verifyOrganism(organism),
): { name: 'baby' | 'hatchling' | 'raptor'; ordinal: number } {
  const dimensions = dimensionStates(organism, report);
  const existing = dimensions.find((entry) => entry.name === dimension);
  if (existing) existing.status = 'active';
  else dimensions.push({ name: dimension, status: 'active', unmeasured: false });
  const name = deriveStageFromEvidence(
    contiguousFrameHeight(organism.frames) + 1,
    dimensions,
  );
  return { name, ordinal: STAGE_ORDINAL[name] };
}

function draftFrame(
  organism: LoadedOrganism,
  dimension: string,
  version: number,
  media: JsonObject,
  createdAt: string,
  report: VerificationReport,
): BodyFrame {
  const seq = organism.frames.length;
  return buildDimensionFrame({
    rappid: organism.document.rappid,
    seq,
    utc: createdAt,
    prev: seq === 0 ? null : organism.frames[seq - 1].payload_hash,
    dimension,
    version,
    stage: projectedStage(organism, dimension, report),
    traits: { ...organism.traits.traitsMilli },
    media,
  });
}

/**
 * One growth proposal for a dimension.
 *
 * The proposal id covers everything that would be written — dimension, kind,
 * evidence and the exact content addresses of every byte — so approving an id
 * approves a specific append and nothing else.
 */
export function buildGrowthProposal(
  organism: LoadedOrganism,
  dimension: string,
  report: VerificationReport = verifyOrganism(organism),
): PendingGrowth {
  if (!isGrowable(dimension)) {
    throw new QuantumRappidError(
      'ungrowable-dimension',
      `no growth path for dimension ${JSON.stringify(dimension)}; known: ${GROWABLE_DIMENSIONS.join(', ')}`,
    );
  }

  const rappid = organism.document.rappid;
  const payloads: Array<{ dimension: string; path: string; bytes: Buffer }> = [];
  let continuation: ContinuationProposal | null = null;
  let assets: AssetRecord[] = [];
  let title: string;
  let summary: string;
  let evidence: string[];
  let frameDimension: string;
  const appendable = dimension === 'sonic';

  if (dimension === 'sonic') {
    const { params, prompt } = sonicContext(organism);
    continuation = proposeContinuation({
      rappid,
      traitsMilli: organism.traits.traitsMilli,
      params,
      prompt,
      engramCursor: organism.document.externalEpisode?.cursor ?? null,
    });
    const bytes = writeMidi([...prompt, ...continuation.continuation], params);
    const path = `assets/autocomplete-${continuation.midiSha256.slice(0, 12)}.mid`;
    assets = [
      {
        path,
        bytes: bytes.length,
        sha256: continuation.midiSha256,
        mediaType: 'application/x-midi',
        durationSeconds: null,
      },
    ];
    payloads.push({ dimension: 'sonic', path, bytes });
    frameDimension = 'sonic';
    title = 'A trait-conditioned continuation of the identity motif';
    summary =
      `${continuation.continuation.length} notes continuing the ${prompt.length}-note MIDI DNA, `
      + `selected from ${continuation.candidateCount} deterministic candidates.`;
    evidence = [
      `provider: ${continuation.provider.name} (${continuation.provider.kind}, not a trained transformer)`,
      `selected candidate ${continuation.selectedCandidate} of ${continuation.candidateCount}`,
      `continuation score ${continuation.scoresMicro.continuation} / 1000000`,
      `standalone score ${continuation.scoresMicro.soundsGood} / 1000000`,
      `rendered midi sha256 ${continuation.midiSha256}`,
    ];
  } else {
    const carried = new Set([
      ...organism.document.dimensions.map((entry) => entry.name),
      ...organism.frames.map((entry) => entry.payload.dimension),
    ]);
    frameDimension =
      ['memory', 'skill', 'sonic', 'device', 'visual', 'capability']
        .find((name) => !carried.has(name))
      ?? 'capability';
    title = `A proposed ${frameDimension} dimension`;
    summary =
      'A lineage- and trait-conditioned preview. It carries no data yet and cannot be appended.';
    evidence = [
      ...report.verifiedAddresses.map((address) => `verified ${address}`),
      `checks passed: ${report.checks.filter((check) => check.status === 'pass').length}`,
    ];
  }

  const version =
    1
    + organism.frames.filter(
      (candidate) => candidate.payload.dimension === frameDimension,
    ).length;
  const media: JsonObject = {};
  const pendingAssets = payloads.map((payload, index) => {
    const asset = assets[index];
    media[
      frameDimension === 'sonic'
        ? 'midi-autocomplete'
        : `asset-${index + 1}`
    ] = mediaRef(payload.bytes, asset.mediaType);
    return { asset, bytes: payload.bytes };
  });
  // Fixed, not a clock read: a preview must predict the same organism twice.
  const frame = draftFrame(
    organism,
    frameDimension,
    version,
    media,
    PREDICTION_INSTANT,
    report,
  );
  const prediction = predict(organism, report, frame, pendingAssets);

  const id = proposalId({
    rappid,
    dimension: frameDimension,
    kind: 'body.dimension',
    version,
    title,
    summary,
    evidence: [...evidence],
    media,
    seq: frame.seq,
    prev: frame.prev,
  });

  return {
    proposal: {
      id,
      rappid,
      dimension: frameDimension,
      title,
      summary,
      predictedStats: prediction.stats,
      predictedStage: prediction.stage,
      evidence,
      assets,
      authoritative: false,
      appendable,
    },
    frameKind: 'body.dimension',
    payloads,
    continuation,
  };
}

export interface GrowthResult {
  rappid: string;
  appended: BodyFrame;
  framePath: string;
  writtenAssets: AssetRecord[];
  summary: QuantumRappidSummary;
  verification: VerificationReport;
}

export interface GrowthOptions {
  /** The frame's recorded time. Defaults to now, in UTC, to the second. */
  createdAt?: string;
}

/**
 * Append an approved proposal.
 *
 * Order matters: verify first, write the bytes, then append the frame that
 * points at them. A frame that names an asset which is not on disk would be a
 * lie the next verification catches — but it is better not to write it at all.
 */
export function growOrganism(
  organism: LoadedOrganism,
  proposalIdentifier: string,
  options: GrowthOptions = {},
): GrowthResult {
  const report = assertVerified(organism);
  const matches = GROWABLE_DIMENSIONS.map((dimension) =>
    buildGrowthProposal(organism, dimension, report),
  ).filter((pending) => pending.proposal.id === proposalIdentifier);

  if (matches.length === 0) {
    throw new QuantumRappidError(
      'unknown-proposal',
      `proposal ${proposalIdentifier} does not match any growth this organism can currently produce; `
        + 'preview it again and approve the new id',
    );
  }
  const pending = matches[0];
  if (!pending.proposal.appendable) {
    throw new QuantumRappidError(
      'proposal-not-appendable',
      'this autocomplete result is a preview only; attach real verified dimension data first',
    );
  }
  const createdAt = options.createdAt ?? formatFrameTime(new Date());
  if (!FRAME_TIME_PATTERN.test(createdAt)) {
    throw new QuantumRappidError(
      'frame-time',
      `createdAt ${createdAt} is not YYYY-MM-DDTHH:MM:SS.mmmZ`,
    );
  }

  const written: AssetRecord[] = pending.payloads.map((payload) =>
    writeDimensionAsset(organism, payload.dimension, payload.path, payload.bytes),
  );
  const version =
    1
    + organism.frames.filter(
      (candidate) =>
        candidate.payload.dimension === pending.proposal.dimension,
    ).length;
  const media: JsonObject = {};
  pending.payloads.forEach((payload, index) => {
    media[
      pending.proposal.dimension === 'sonic'
        ? 'midi-autocomplete'
        : `asset-${index + 1}`
    ] = mediaRef(payload.bytes, written[index].mediaType);
  });
  const frame = draftFrame(
    organism,
    pending.proposal.dimension,
    version,
    media,
    createdAt,
    report,
  );
  const framePath = appendBodyFrame(organism, frame);
  const verification = verifyOrganism(organism);

  return {
    rappid: organism.document.rappid,
    appended: frame,
    framePath,
    writtenAssets: written,
    summary: summarize(organism, verification),
    verification,
  };
}

File: repos/openrappter/python/openrappter/rappids/growth.py
"""Growth: append a verified body frame, never mint a second identity.

Mirrors ``typescript/src/rappids/growth.ts``.

A proposal is not organism state. It is a preview with a content-addressed id,
and the id is what makes approval mean something: ``grow_organism``
re-derives the proposal from the organism itself and refuses anything whose id
does not match, so an approved preview cannot be swapped for a different
payload between the preview and the append.

Appending never touches the RAPPID. A creature that grows from baby to Raptor
is the same creature the whole way -- only a true child gets a new identity,
and it gets an explicit parent pointer with it.
"""

from __future__ import annotations

from datetime import datetime, timezone
from dataclasses import replace
from typing import Any, Dict, List, Optional, Sequence, Tuple

from .autocomplete import propose_continuation
from .canonical import (
    PROPOSAL_DOMAIN,
    RAPP_EGG_DOMAIN,
    canonical_json,
    domain_digest,
    rapp_hb,
)
from .midi import build_dna_prompt, sonic_parameters, write_midi
from .stats import (
    contiguous_frame_height,
    derive_stage,
    derive_stage_from_evidence,
    derive_stats,
    dimension_states,
    summarize,
)
from .store import (
    FRAME_TIME_PATTERN,
    append_body_frame,
    build_dimension_frame,
    format_frame_time,
    media_ref,
    write_dimension_asset,
)
from .types import (
    AssetRecord,
    AssetVerification,
    BodyFrame,
    ContinuationProposal,
    DimensionState,
    GrowthProposal,
    LoadedOrganism,
    MusicalParameters,
    Note,
    QuantumRappidError,
    VerificationReport,
)
from .verify import assert_verified, verify_organism

#: Growth requests this runtime knows how to answer. Anything else is refused.
GROWABLE_DIMENSIONS = ("sonic", "stats")

#: The instant a prediction is drafted against. Frame timestamps are
#: fixed-width, so any value of the right shape produces a frame of the same
#: canonical length -- which is what lets a preview quote exact bytes for a
#: frame that has not been written yet.
PREDICTION_INSTANT = "1970-01-01T00:00:00.000Z"

#: The lifecycle ladder as a frame records it: a name and its ordinal.
STAGE_ORDINAL = {"baby": 0, "hatchling": 1, "raptor": 2}

#: Families a stats-only preview may propose, in the order it prefers them.
PROPOSABLE_DIMENSIONS = ("memory", "skill", "sonic", "device", "visual", "capability")


class PendingGrowth:
    """A proposal plus the bytes it would write, held together until approval."""

    def __init__(
        self,
        proposal: GrowthProposal,
        frame_kind: str,
        payloads: Sequence[Tuple[str, str, bytes]],
        continuation: Optional[ContinuationProposal],
    ) -> None:
        self.proposal = proposal
        self.frame_kind = frame_kind
        #: ``(dimension, path, bytes)`` for every file the append would write.
        self.payloads = list(payloads)
        self.continuation = continuation


def sonic_context(organism: LoadedOrganism) -> Tuple[MusicalParameters, List[Note]]:
    """The prompt and key an organism sings in, recorded if it has a profile."""
    params = (
        organism.sonic.musical
        if organism.sonic is not None
        else sonic_parameters(organism.document.rappid, organism.traits.birth_traits_milli)
    )
    prompt = (
        list(organism.sonic.prompt)
        if organism.sonic is not None
        else build_dna_prompt(
            organism.document.rappid, organism.traits.birth_traits_milli, params
        )
    )
    return params, prompt


def _proposal_id(body: Dict[str, Any]) -> str:
    return domain_digest(PROPOSAL_DOMAIN, canonical_json(body))


def _pending_verification(
    dimension: str, asset: AssetRecord, payload: bytes
) -> AssetVerification:
    """A verified asset record for bytes that are in hand but not yet on disk."""
    return AssetVerification(
        dimension=dimension,
        path=asset.path,
        status="verified",
        address_space=RAPP_EGG_DOMAIN,
        address_hash=rapp_hb(RAPP_EGG_DOMAIN, payload),
        expected_bytes=asset.bytes,
        actual_bytes=asset.bytes,
        expected_sha256=asset.sha256,
        actual_sha256=asset.sha256,
        media_type=asset.media_type,
    )


def projected_stage(
    organism: LoadedOrganism,
    dimension: str,
    report: Optional[VerificationReport] = None,
) -> Dict[str, Any]:
    """The stage this frame would record, read off the body it would create.

    The dimension being appended counts as active because the frame is what
    makes it active; everything else is the organism's current evidence.
    """
    if report is None:
        report = verify_organism(organism)
    dimensions: List[DimensionState] = []
    seen = False
    for entry in dimension_states(organism, report):
        if entry.name == dimension:
            seen = True
            dimensions.append(replace(entry, status="active"))
        else:
            dimensions.append(entry)
    if not seen:
        dimensions.append(
            DimensionState(name=dimension, status="active", media_types=[], unmeasured=False)
        )
    name = derive_stage_from_evidence(
        contiguous_frame_height(organism.frames) + 1, dimensions
    )
    return {"name": name, "ordinal": STAGE_ORDINAL[name]}


def _draft_frame(
    organism: LoadedOrganism,
    dimension: str,
    version: int,
    media: Dict[str, Any],
    created_at: str,
    report: VerificationReport,
) -> BodyFrame:
    """Build the frame a proposal would append.

    ``created_at`` is an input rather than a clock read, because the frame hash
    covers it: a proposal id that changed every second could never be approved.
    """
    seq = len(organism.frames)
    return build_dimension_frame(
        rappid=organism.document.rappid,
        seq=seq,
        utc=created_at,
        prev=None if seq == 0 else organism.frames[seq - 1].payload_hash,
        dimension=dimension,
        version=version,
        stage=projected_stage(organism, dimension, report),
        traits=dict(organism.traits.traits_milli),
        media=media,
    )


def _predict(
    organism: LoadedOrganism,
    report: VerificationReport,
    frame: BodyFrame,
    payloads: Sequence[Tuple[AssetRecord, bytes]],
):
    """What the organism would look like after this frame, without writing it.

    The prediction runs the real derivation over a copy rather than a parallel
    "estimate" path: a preview computed differently from the thing it previews
    is how a habitat starts lying to its operator.
    """
    dimension = str(frame.payload.get("dimension"))
    projected = LoadedOrganism(
        directory=organism.directory,
        document=organism.document,
        traits=organism.traits,
        sonic=organism.sonic,
        frames=[*organism.frames, frame],
    )
    projected_report = VerificationReport(
        rappid=report.rappid,
        ok=report.ok,
        checks=report.checks,
        assets=[
            *report.assets,
            *[
                _pending_verification(dimension, asset, payload)
                for asset, payload in payloads
            ],
        ],
        verified_addresses=sorted(
            {
                *report.verified_addresses,
                *[
                    f"{RAPP_EGG_DOMAIN}:{rapp_hb(RAPP_EGG_DOMAIN, payload)}"
                    for _asset, payload in payloads
                ],
            }
        ),
    )
    dimensions = dimension_states(projected, projected_report)
    stats = derive_stats(projected, projected_report, dimensions)
    return stats, derive_stage(stats, dimensions)


def build_growth_proposal(
    organism: LoadedOrganism, dimension: str, report: Optional[VerificationReport] = None
) -> PendingGrowth:
    """One growth proposal for a dimension.

    The proposal id covers everything that would be written -- dimension, kind,
    evidence and the exact content addresses of every byte -- so approving an
    id approves a specific append and nothing else.
    """
    if dimension not in GROWABLE_DIMENSIONS:
        raise QuantumRappidError(
            "ungrowable-dimension",
            f"no growth path for dimension {dimension!r}; "
            f"known: {', '.join(GROWABLE_DIMENSIONS)}",
        )
    if report is None:
        report = verify_organism(organism)

    rappid = organism.document.rappid
    payloads: List[Tuple[str, str, bytes]] = []
    continuation: Optional[ContinuationProposal] = None
    assets: List[AssetRecord] = []
    appendable = dimension == "sonic"

    if dimension == "sonic":
        params, prompt = sonic_context(organism)
        cursor = (
            organism.document.external_episode.cursor
            if organism.document.external_episode is not None
            else None
        )
        continuation = propose_continuation(
            rappid, organism.traits.traits_milli, params, prompt, cursor
        )
        payload = write_midi([*prompt, *continuation.continuation], params)
        path = f"assets/autocomplete-{continuation.midi_sha256[:12]}.mid"
        assets = [
            AssetRecord(
                path=path,
                bytes=len(payload),
                sha256=continuation.midi_sha256,
                media_type="application/x-midi",
                duration_seconds=None,
            )
        ]
        payloads.append(("sonic", path, payload))
        frame_dimension = "sonic"
        title = "A trait-conditioned continuation of the identity motif"
        summary = (
            f"{len(continuation.continuation)} notes continuing the {len(prompt)}-note MIDI DNA, "
            f"selected from {continuation.candidate_count} deterministic candidates."
        )
        evidence = [
            f"provider: {continuation.provider['name']} "
            f"({continuation.provider['kind']}, not a trained transformer)",
            f"selected candidate {continuation.selected_candidate} of {continuation.candidate_count}",
            f"continuation score {continuation.scores_micro.continuation} / 1000000",
            f"standalone score {continuation.scores_micro.sounds_good} / 1000000",
            f"rendered midi sha256 {continuation.midi_sha256}",
        ]
    else:
        carried = {entry.name for entry in organism.document.dimensions} | {
            str(entry.payload.get("dimension")) for entry in organism.frames
        }
        frame_dimension = next(
            (name for name in PROPOSABLE_DIMENSIONS if name not in carried), "capability"
        )
        title = f"A proposed {frame_dimension} dimension"
        summary = (
            "A lineage- and trait-conditioned preview. It carries no data yet and "
            "cannot be appended."
        )
        evidence = [
            *[f"verified {address}" for address in report.verified_addresses],
            f"checks passed: {sum(1 for check in report.checks if check.status == 'pass')}",
        ]

    version = 1 + sum(
        1
        for candidate in organism.frames
        if str(candidate.payload.get("dimension")) == frame_dimension
    )
    media: Dict[str, Any] = {}
    pending_assets: List[Tuple[AssetRecord, bytes]] = []
    for index, (_dimension, _path, payload_bytes) in enumerate(payloads):
        asset = assets[index]
        role = "midi-autocomplete" if frame_dimension == "sonic" else f"asset-{index + 1}"
        media[role] = media_ref(payload_bytes, asset.media_type)
        pending_assets.append((asset, payload_bytes))

    # Fixed, not a clock read: a preview must predict the same organism twice.
    frame = _draft_frame(
        organism, frame_dimension, version, media, PREDICTION_INSTANT, report
    )
    stats, stage = _predict(organism, report, frame, pending_assets)

    proposal_id = _proposal_id(
        {
            "rappid": rappid,
            "dimension": frame_dimension,
            "kind": "body.dimension",
            "version": version,
            "title": title,
            "summary": summary,
            "evidence": list(evidence),
            "media": media,
            "seq": frame.seq,
            "prev": frame.prev,
        }
    )

    return PendingGrowth(
        proposal=GrowthProposal(
            id=proposal_id,
            rappid=rappid,
            dimension=frame_dimension,
            title=title,
            summary=summary,
            predicted_stats=stats,
            predicted_stage=stage,
            evidence=list(evidence),
            assets=assets,
            appendable=appendable,
        ),
        frame_kind="body.dimension",
        payloads=payloads,
        continuation=continuation,
    )


def grow_organism(
    organism: LoadedOrganism, proposal_id: str, created_at: Optional[str] = None
) -> Dict[str, Any]:
    """Append an approved proposal.

    Order matters: verify first, write the bytes, then append the frame that
    points at them. A frame that names an asset which is not on disk would be a
    lie the next verification catches -- but it is better not to write it.
    """
    report = assert_verified(organism)
    matches = [
        pending
        for pending in (
            build_growth_proposal(organism, dimension, report)
            for dimension in GROWABLE_DIMENSIONS
        )
        if pending.proposal.id == proposal_id
    ]
    if not matches:
        raise QuantumRappidError(
            "unknown-proposal",
            f"proposal {proposal_id} does not match any growth this organism can currently "
            "produce; preview it again and approve the new id",
        )
    pending = matches[0]
    if not pending.proposal.appendable:
        raise QuantumRappidError(
            "proposal-not-appendable",
            "this autocomplete result is a preview only; attach real verified "
            "dimension data first",
        )
    stamp = created_at if created_at is not None else format_frame_time(datetime.now(timezone.utc))
    if not FRAME_TIME_PATTERN.fullmatch(stamp):
        raise QuantumRappidError(
            "frame-time",
            f"created_at {stamp} is not YYYY-MM-DDTHH:MM:SS.mmmZ",
        )

    written = [
        write_dimension_asset(organism, dimension, path, payload)
        for dimension, path, payload in pending.payloads
    ]
    version = 1 + sum(
        1
        for candidate in organism.frames
        if str(candidate.payload.get("dimension")) == pending.proposal.dimension
    )
    media: Dict[str, Any] = {}
    for index, (_dimension, _path, payload_bytes) in enumerate(pending.payloads):
        role = (
            "midi-autocomplete"
            if pending.proposal.dimension == "sonic"
            else f"asset-{index + 1}"
        )
        media[role] = media_ref(payload_bytes, written[index].media_type)
    frame = _draft_frame(
        organism, pending.proposal.dimension, version, media, stamp, report
    )
    frame_path = append_body_frame(organism, frame)
    verification = verify_organism(organism)

    return {
        "rappid": organism.document.rappid,
        "appended": frame.to_wire(),
        "framePath": frame_path,
        "writtenAssets": [asset.to_wire() for asset in written],
        "summary": summarize(organism, verification),
        "verification": verification.to_wire(),
    }
